File: HeartBeatSystemDemo/HeartBeatSystemDemo/src/HeartBeatSystem.java
/**
 * This code is a DEMO for HeartBeatSystem
 * Which is an idea that come out from Sandoleathy
 * Date: 2023/9/11
 */

import java.util.LinkedList;
import java.util.Queue;
import java.util.Random;
import java.util.Stack;

public class HeartBeatSystem {
    private HeartBeatStatus bodyState;
    private int heartBeatRate;
    private Queue<BPMChange> changeQueue;
    private LinkedList<Drags> dragList;
    public final static int DEAD_LINE_HIGHEST = 220;
    public final static int DEAD_LINE_LOWEST = 45;
    public HeartBeatSystem(){
        bodyState = HeartBeatStatus.RESTING;
        heartBeatRate = 60;
        changeQueue = new LinkedList<BPMChange>();
        dragList = new LinkedList<Drags>();
    }
    //get shot or take damage
    public void takeHit(){
        changeQueue.offer(new BPMChange(500,0,10));
    }
    public void jump(){
        changeQueue.offer(new BPMChange(200,40,5));
    }
    public void slowRun(){
        changeQueue.offer(new BPMChange(150,40,6));
    }
    public void deepBreathe(){
        changeQueue.offer(new BPMChange(220,70,-10));
    }
    public void fastRun(){
        changeQueue.offer(new BPMChange(200,40,16));
    }
    public void heavyBleed(){
        changeQueue.offer(new BPMChange(500,0,5));
    }
    public void fall(){
        changeQueue.offer(new BPMChange(500,0,10));
    }

    /**
     * Make sure body's heartbeat will randomly float
     */
    public void randomChangeHeartBeat(){
        if(changeQueue.size() < 3){
            Random random = new Random();
            if(bodyState == HeartBeatStatus.RESTING){
                if(heartBeatRate > 80){
                    changeQueue.offer(new BPMChange(500,0,-(random.nextInt(10))));
                }else if(heartBeatRate < 60){
                    changeQueue.offer(new BPMChange(500,0,random.nextInt(10)));
                }else{
                    changeQueue.offer(new BPMChange(500,0, (random.nextInt(10)) - 5 ));
                }
            }
            else if(bodyState == HeartBeatStatus.RAGE){
                if(heartBeatRate > 160){
                    changeQueue.offer(new BPMChange(500,0,-(random.nextInt(10))));
                }else if(heartBeatRate < 130){
                    changeQueue.offer(new BPMChange(500,0,random.nextInt(10)));
                }else{
                    changeQueue.offer(new BPMChange(500,0, (random.nextInt(10)) - 5 ));
                }
            }
            else if(bodyState == HeartBeatStatus.NERVOUS){
                if(heartBeatRate > 110){
                    changeQueue.offer(new BPMChange(500,0,-(random.nextInt(10))));
                }else if(heartBeatRate < 80){
                    changeQueue.offer(new BPMChange(500,0,random.nextInt(10)));
                }else{
                    changeQueue.offer(new BPMChange(500,0, (random.nextInt(10)) - 5 ));
                }
            }
        }
    }
    public void useDrag(Drags drag){
        dragList.add(drag);
        if(drag.getType() == DragType.ADRENALINE){//using adrenaline
            changeBodyState(HeartBeatStatus.RAGE);
            changeQueue.offer(new BPMChange(500,0,25));
            changeQueue.offer(new BPMChange(500,0,25));
        }
        else if(drag.getType() == DragType.TRANQUILIZER){//use tranquilizer
            changeBodyState(HeartBeatStatus.RESTING);
            changeQueue.offer(new BPMChange(500,0,-25));
            changeQueue.offer(new BPMChange(500,0,-25));
        }
    }

    /**
     * pull a value from BPM changeList
     * add it to HeartBeat
     * also check BPM and update body state
     */
    public void updateHeartBeat(){
        if(changeQueue.isEmpty()){
            return;
        }
        BPMChange change = changeQueue.poll();
        if(heartBeatRate < change.getUpLimit() && heartBeatRate > change.getDownLimit()){
            heartBeatRate = heartBeatRate + change.getBPMChange();
        }
        if(heartBeatRate <= 80){
            bodyState = HeartBeatStatus.RESTING;
        }
        else if(heartBeatRate <= 110){
            bodyState = HeartBeatStatus.NERVOUS;
        }
        else {
            bodyState = HeartBeatStatus.RAGE;
        }
    }
    private void changeBodyState(HeartBeatStatus type){
        bodyState = type;
    }
    public int getHeartBeatRate(){
        return heartBeatRate;
    }
    public Queue<BPMChange> getChangeQueue(){
        return changeQueue;
    }
    public HeartBeatStatus getBodyState(){
        return bodyState;
    }

}

File: UnityProjet/scripts/HeartBeatSystem/models/HeartBeatSystem.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using UnityEngine;

public class HeartBeatSystem : MonoBehaviour
{
    public int HeartBeatRate;
    private HeartBeatStatus BodyState;
    public const int DEAD_LINE_HIGHEST = 220;
    public const int DEAD_LINE_LOWEST = 45;
    private Queue<BPMChange> ChangeQueue;
    private LinkedList<IDrags> DragList;
    private Thread DragThread1 = null;
    private Thread HeartBeatThread1 = null;
    private HeartBeatThread HBSThread;
    private DragThread DragTh;

    // Start is called before the first frame update
    // Start function also use as a constructor
    void Start()
    {
        BodyState = HeartBeatStatus.RESTING;
        HeartBeatRate = 60;
        ChangeQueue = new Queue<BPMChange>();
        DragList = new LinkedList<IDrags>();
        //init hearbeat thread and start it
        HBSThread = new HeartBeatThread(this);
        HeartBeatThread1 = new Thread(HBSThread.Run);
        HeartBeatThread1.IsBackground = true;
        HeartBeatThread1.Start();
        DragTh = new DragThread(this);
        Debug.Log("initialized success process start");
    }

    // Update is called once per frame
    // You can consider Update function as a Main function
    void Update()
    {
        
    }
    private void OnApplicationPause(bool pause)
    {
        if (DragTh == null || HBSThread == null)
        {
            return;
        }
        if (pause)
        {
            Debug.Log("Pause");
            DragTh.Pause();
            HBSThread.Pause();
        }
        else
        {
            DragTh.UnPause();
            HBSThread.UnPause();
        }
        
    }
    // When game quit shutdown threads
    private void OnApplicationQuit()
    {
        if(HeartBeatThread1 != null)
        {
            HeartBeatThread1.Abort();
        }
        if(DragThread1 != null)
        {
            DragThread1.Abort();
        }
    }
    public void TakeHit()
    {
        ChangeQueue.Enqueue( new BPMChange( 500 , 0 , 10 ) );
    }
    public void Jump()
    {
        ChangeQueue.Enqueue(new BPMChange(200, 40, 5));
    }
    public void SlowRun()
    {
        ChangeQueue.Enqueue(new BPMChange(150, 40, 16));
    }
    public void DeepBreathe()
    {
        ChangeQueue.Enqueue(new BPMChange(220, 70, -10));
    }
    public void FastRun()
    {
        ChangeQueue.Enqueue(new BPMChange(200, 40, 16));
    }
    public void HeavyBleed()
    {
        ChangeQueue.Enqueue(new BPMChange(500, 0, 5));
    }
    public void Fall()
    {
        ChangeQueue.Enqueue(new BPMChange(500, 0, 10));
    }
    private void ChangeQueueAdd(BPMChange change)
    {
        /*
         * 使用了和Java相同的同步锁来防止异步操作
         */
        lock (this.ChangeQueue)
        {
            ChangeQueue.Enqueue(change);
        }
    }
    public void useDrag(IDrags drag)
    {
        DragList.AddFirst(drag);
        if (drag.GetType() == DragTypes.ADRENALINE)
        {//using adrenaline
            ChangeBodyState(HeartBeatStatus.RAGE);
            ChangeQueueAdd(new BPMChange(500, 0, 25));
            ChangeQueueAdd(new BPMChange(500, 0, 25));
        }
        else if (drag.GetType() == DragTypes.TRANQUILIZER)
        {//use tranquilizer
            ChangeBodyState(HeartBeatStatus.RESTING);
            ChangeQueueAdd(new BPMChange(500, 0, -25));
            ChangeQueueAdd(new BPMChange(500, 0, -25));
        }
        if(DragThread1 != null)
        {
            return;
        }
        DragThread1= new Thread(DragTh.Run);
        DragThread1.IsBackground = true;
        DragThread1.Start();
        /*
        if (thread != null)
        {
            return;
        }
        thread = new DragThread(this);
        new Thread(thread).start();
        */
    }
    /**
     * Make sure body's heartbeat will randomly float
     */
    public void RandomrandomChangeHeartBeat()
    {
        if(ChangeQueue.Count >= 3)
        {
            return;
        }
        System.Random ran = new System.Random();
        int i = ran.Next(10);
        if(BodyState == HeartBeatStatus.RESTING)
        {
            if (HeartBeatRate > 80)
            {
                ChangeQueueAdd( new BPMChange( 500, 0, -i ) );
            }
            else if (HeartBeatRate < 60)
            {
                ChangeQueueAdd( new BPMChange( 500, 0, i ) );
            }
            else
            {
                ChangeQueueAdd(new BPMChange(500, 0, i-5 ) );
            }
        }else if (BodyState == HeartBeatStatus.RAGE)
        {
            if (HeartBeatRate > 160)
            {
                ChangeQueueAdd(new BPMChange(500, 0, -i ));
            }
            else if (HeartBeatRate < 130)
            {
                ChangeQueueAdd(new BPMChange(500, 0, i ));
            }
            else
            {
                ChangeQueueAdd(new BPMChange(500, 0, i-5 ));
            }
        }else if (BodyState == HeartBeatStatus.NERVOUS)
        {
            if (HeartBeatRate > 110)
            {
                ChangeQueueAdd(new BPMChange(500, 0, -i ));
            }
            else if (HeartBeatRate < 80)
            {
                ChangeQueueAdd(new BPMChange(500, 0, i ));
            }
            else
            {
                ChangeQueueAdd(new BPMChange(500, 0, i-5 ));
            }
        }
    }
    public void UpdateHeartBeat()
    {
        Debug.Log("Update Heartbeat");
        if(ChangeQueue.Count == 0)
        {
            return;
        }
        BPMChange Change = ChangeQueue.Dequeue();
        if (HeartBeatRate < Change.upLimit && HeartBeatRate > Change.downLimit)
        {
            HeartBeatRate = HeartBeatRate + Change.changeValue;
        }
        if (HeartBeatRate <= 80)
        {
            BodyState = HeartBeatStatus.RESTING;
        }
        else if (HeartBeatRate <= 110)
        {
            BodyState = HeartBeatStatus.NERVOUS;
        }
        else
        {
            BodyState = HeartBeatStatus.RAGE;
        }
    }
    public void RemoveDrag(IDrags d)
    {
        Debug.Log("Remove Drag");
        DragList.Remove(d);
    }
    public void ResetHeartBeatRate()
    {
        Debug.Log("Reset HeartBeat Rate");
        this.HeartBeatRate = 60;
    }
    public void ChangeBodyState( HeartBeatStatus h )
    {
        this.BodyState = h;
    }
    public int GetHeartBeatRate()
    {
        return HeartBeatRate;
    }
    public HeartBeatStatus GetBodyState()
    {
        return BodyState;
    }
    public Queue<BPMChange> GetChangeQueue()
    {
        return ChangeQueue;
    }
    public LinkedList<IDrags> GetDragList()
    {
        return DragList;
    }
}
